ipv6/uint128.go

```diff
@@ -0,0 +1,207 @@
+*/
+
+package ipv6
+
+import "math/bits"
+
+type uint128 struct {
+	high, low uint64
+}
+
+// Uint128FromUint64 returns a uint128 from its two 64 bit unsigned representations
+func Uint128FromUint64(high uint64, low uint64) uint128 {
+	return uint128{high, low}
+}
+
+// Uint128FromBytes returns the the uint128 converted from a slice of 16 bytes
+func Uint128FromBytes(s []byte) uint128 {
+	return uint128{
+		high: uint64(s[0])<<56 | uint64(s[1])<<48 | uint64(s[2])<<40 | uint64(s[3])<<32 | uint64(s[4])<<24 | uint64(s[5])<<16 | uint64(s[6])<<8 | uint64(s[7]),
+		low:  uint64(s[8])<<56 | uint64(s[9])<<48 | uint64(s[10])<<40 | uint64(s[11])<<32 | uint64(s[12])<<24 | uint64(s[13])<<16 | uint64(s[14])<<8 | uint64(s[15]),
+	}
+}
+
+// MaxUint128 is the maximum integer that can be stored in a uint128, "all ones"
+func MaxUint128() uint128 {
+	return uint128{^uint64(0), ^uint64(0)}
+}
+
+// OnesCount128 returns the number of one bits ("population count") in x.
+func OnesCount128(x uint128) int {
+	return bits.OnesCount64(x.high) + bits.OnesCount64(x.low)
+}
+
+// LeadingZeros128 returns the number of leading zero bits in x; the result is 128 for x == 0.
+func LeadingZeros128(x uint128) int {
+	leadingZeros := bits.LeadingZeros64(x.high)
+	if leadingZeros == 64 {
+		leadingZeros += bits.LeadingZeros64(x.low)
+	}
+	return leadingZeros
+}
+
+// ToBytes returns a slice representation of 16 bytes of the uint128
+func (me uint128) ToBytes() []byte {
+	bytes := []byte{
+		byte(0xff & (me.high >> 56)),
+		byte(0xff & (me.high >> 48)),
+		byte(0xff & (me.high >> 40)),
+		byte(0xff & (me.high >> 32)),
+		byte(0xff & (me.high >> 24)),
+		byte(0xff & (me.high >> 16)),
+		byte(0xff & (me.high >> 8)),
+		byte(0xff & me.high),
+		byte(0xff & (me.low >> 56)),
+		byte(0xff & (me.low >> 48)),
+		byte(0xff & (me.low >> 40)),
+		byte(0xff & (me.low >> 32)),
+		byte(0xff & (me.low >> 24)),
+		byte(0xff & (me.low >> 16)),
+		byte(0xff & (me.low >> 8)),
+		byte(0xff & me.low),
+	}
+	return bytes
+}
+
+// Uint64 returns the address as two uint64
+func (me uint128) Uint64() (uint64, uint64) {
+	return me.high, me.low
+}
+
+// Equal reports whether this uint128 is the same as other
+func (me uint128) Equal(other uint128) bool {
+	return me == other
+}
+
+// Compare returns comparison of two uint128s and returns:
+//  O if equal
+// -1 if me is less than other
+//  1 if me is greater than other
+func (me uint128) Compare(other uint128) int {
+	if me == other {
+		return 0
+	} else if me.high < other.high || (me.high == other.high && me.low < other.low) {
+		return -1
+	} else {
+		return 1
+	}
+}
+
+// IsZero returns whether uint128 equals zero
+func (me uint128) IsZero() bool {
+	return me.low == 0 && me.high == 0
+}
+
+// SubtractUint64 returns difference of y (uint64) from x (uint128)
+func SubtractUint64(x uint128, y uint64) uint128 {
```

Comment:
I felt that in the case of a receiver, i.e address.ui.SubtractUint64(other) this would generally be seen as it having modified the existing address.ui. My intent was to return a whole new uint128 which I felt that doing it in the form SubtractUint64(x,y) over ui.SubtractUint64(other) is of better convention. Much like how append() works.

Reply:
You're right that we do not want to modify the original value. The type must be immutable.

The thing is, it already is immutable even without this distinction. Since this method would have a non-pointer receiver -- like the other existing methods on this type -- and has no pointers to other mutable values, those aspects are already guaranteed. There is no choice but to compute and return a new value.

You can decide. I just wanted to point out that it is only a matter of perception and doesn't actually provide any guarantees beyond what is already there.

As an aside: I spent some time questioning my earlier use of pointer receivers in the exposed public API and if you rebase, you'll notice that I've already removed all of them. Only trieNode and setNode still have a pointer receiver and that was because 1) they are nodes in a trie structure linked together with pointers and 2) I felt their struct size was a bit large and I wondered if pointer receivers would optimize it a bit. They both are intended to behave like immutable things in most, if not all ways.

As I'm writing this, I'm wondering about my assumptions. Could those could be implemented the other way and what would the performance impact be? Since it isn't part of the exposed API, I may leave that for a later exercise. It certainly would make it easier to ensure that a node cannot be modified once it exists.